x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizerTests.java

```diff
@@ -1226,11 +1230,268 @@ public boolean isIndexed(FieldAttribute.FieldName field) {
+        assertTrue(esRelation.output().contains(fieldAttr));
+    }
+
+    public void testVectorFunctionsInStats() {
+        assumeTrue("requires similarity functions", EsqlCapabilities.Cap.VECTOR_SIMILARITY_FUNCTIONS_PUSHDOWN.isEnabled());
+        String query = """
+            from test_all
+            | stats count(*) where v_dot_product(dense_vector, [1.0, 2.0, 3.0]) > 0.5
+            """;
+
+        LogicalPlan plan = localPlan(plan(query, allTypesAnalyzer), TEST_SEARCH_STATS);
+
+        // Limit[1000[INTEGER],false,false]
+        var limit = as(plan, Limit.class);
+        assertThat(limit.limit().fold(FoldContext.small()), equalTo(1000));
+
+        // Aggregate[[],[COUNT(*[KEYWORD],true[BOOLEAN]) AS count(*) where v_dot_product(dense_vector, [1.0, 2.0, 3.0]) > 0.5#3]]
+        var aggregate = as(limit.child(), Aggregate.class);
+        assertThat(aggregate.groupings(), hasSize(0));
+        assertThat(aggregate.aggregates(), hasSize(1));
+
+        // Check the Count aggregate with filter
+        var countAlias = as(aggregate.aggregates().getFirst(), Alias.class);
+        var count = as(countAlias.child(), org.elasticsearch.xpack.esql.expression.function.aggregate.Count.class);
+
+        // Check the filter on the Count aggregate
+        assertThat(count.filter(), equalTo(Literal.TRUE));
+
+        // Filter[$$dense_vector$DotProduct$26{f}#26 > 0.5[DOUBLE]]
+        var filter = as(aggregate.child(), Filter.class);
+        var filterCondition = as(filter.condition(), GreaterThan.class);
+
+        // Check left side is the replaced field attribute
+        var fieldAttr = as(filterCondition.left(), FieldAttribute.class);
+        assertThat(fieldAttr.fieldName().string(), equalTo("dense_vector"));
+        assertThat(fieldAttr.name(), startsWith("$$dense_vector$DotProduct"));
+        var field = as(fieldAttr.field(), FunctionEsField.class);
+        var blockLoaderFunctionConfig = as(field.functionConfig(), DenseVectorFieldMapper.VectorSimilarityFunctionConfig.class);
+        assertThat(blockLoaderFunctionConfig.similarityFunction(), is(DotProduct.SIMILARITY_FUNCTION));
+        assertThat(blockLoaderFunctionConfig.vector(), equalTo(new float[] { 1.0f, 2.0f, 3.0f }));
+
+        // Verify the filter condition matches the aggregate filter
+        var filterFieldAttr = as(filterCondition.left(), FieldAttribute.class);
+        assertThat(filterFieldAttr, is(fieldAttr));
+
+        // EsRelation[test_all][$$dense_vector$DotProduct$26{f}#26, !alias_integer, ..]
+        var esRelation = as(filter.child(), EsRelation.class);
+        assertTrue(esRelation.output().contains(filterFieldAttr));
+    }
+
+    public void testVectorFunctionsUpdateIntermediateProjections() {
+        assumeTrue("requires similarity functions", EsqlCapabilities.Cap.VECTOR_SIMILARITY_FUNCTIONS_PUSHDOWN.isEnabled());
+        String query = """
+            from test_all
+            | keep *
+            | mv_expand keyword
```

Comment:
nit: we'll implement pushdown past project for mv expand very very soon, making this test a bit pointless. We could parse the plan and manually place a custom unary plan node after the projection, if we want to make this robust.